### PR TITLE
Consolidate typeahead

### PR DESCRIPTION
**Pull request by jmcarp**
Branched off `styling-calendar`.

To be determined:
* [x] Should all typeahead filters support search by ID and free text?
* [x] How to clarify the behavior of each filter (ID vs free text)

cc @noahmanger @LindsayYoung @jenniferthibault

### CODE REVIEW

**Comment by jmcarp** [comment]
Assigning this to @noahmanger to coordinate decision-making around filter behavior, update the issue once choices get made, and possibly review changes once everything's implemented.

**Comment by noahmanger** [comment]
Thanks for getting this started. I think there's a pretty strong argument that we'll want certain filters to *only* accept typeahead, so I wanted to try to figure out how the two types would be distinguished. 

typeahead

This is a pretty simple solution. I imagine there's maybe more we could do to differentiate, but thought we could start here.

Of course, if we figure that some of the fields could actually accept free text, then this isn't a problem, but worth exploring either way.

@jenniferthibault what do you think?

**Comment by jenniferthibault** [comment]
Before I amend this with a small behavioral addition, I want to make sure I'm not off-base. @noahmanger is it possible that for a type 1 search, a user could start typing "bill", see all of the below typeahead options, and hit enter to search for "bill" as plaintext?

Or is the behavior that you can only enter a basic text string IF (and only if) there is no matching or suggested typeahead options?

**Comment by noahmanger** [comment]
The first one: you can hit enter and search just text (or click the bottom part of the dropdown).

**Comment by jenniferthibault** [comment]
The functionality is totally in line with how I expected the behavior to work, but I'd like to suggest a few text changes to make language more concise.

`Matching committees` >> `Committees`
`Search for all contributors named: “bill” ` >> `"bill"` ; and move this to the first option under the search box, instead of last. This is a more common pattern, and I think logical to keep them visually grouped.
Defer entirely to @emileighoutlaw 's judgement on microcopy here.

And some styling tweaks:
- the darker shade should act as a highlight, and default to the first suggested option, whether that's `"word"` or option in a typeahead, so that when a user hits enter, it selects what is highlighted as the query
- is it possible to highlight (or in this design, make bold) the part of the word matching the text entry in the results? If bold isn't distinct enough, we could consider a text highlight.
- try not to use italics
- small indent on the list of committees under the matching results header
- shifting placement of the search icon to the end of the bar so that search text can remain aligned with results

Here's what that looks like, with Noah's original above for easy comparison: (better viewing experience if you open in a new tab)
typeahead-type 1
typeahead-type 2

**Comment by emileighoutlaw** [comment]
I love Jen's microcopy suggestions. 

I have one question about the Committee ID field. I love the note added for clarity, but even with it, `Committee ID` only looks like a place to enter a committee ID to me. What do you think about something more like `Search for a committee's ID` or `Find a committee's ID` ?

**Comment by noahmanger** [comment]
@emileighoutlaw great question, and that's exactly what we're struggling with. That sounds like it could work.

@jenniferthibault love the modifications. Only question I have is if the search icon button on the right is meant as a button and if so, what does it do?

**Comment by jenniferthibault** [comment]
Nope, not meant as a button, just as an icon that brings more attention to the field. It wouldn't have any interaction. Similar to Google's search field here, all the interaction happens through typeahead or when the user hits enter to commit the term to search.

https://www.google.com/#q=law+and+order
screen shot 2016-01-31 at 8 05 09 pm

**Comment by noahmanger** [comment]
Cool cool. So, even in the google case it's still a button, but I guess it doesn't really serve a function other than just being there as a safety net for folks. 

@jmcarp I'd say were good to go on continuing implementation based on Jen's mocks above. The latest work on the new datatable columns and filters suggests that having a way to require a specific candidate / committee ID will be helpful, so I'd like to at least get this to a point where can do some quick and dirty usability testing and see if people get too hung up before we try anything else.

**Comment by jenniferthibault** [comment]
On the search bar, right! A dummy/safety button of sorts. Same as we're using here:
screen shot 2016-02-01 at 10 11 42 pm

I was thinking of this existing on our site, but should have added this screenshot the first time around.

**Comment by jmcarp** [comment]
I pushed some updates to fec-style/consolidate-typeahead and openfec-web-app/consolidate-typeahead that behave like @jenniferthibault's designs (I haven't addressed anything text-related). To test it out, verify that the contributor filter allows text search on the receipts data table page, but the committee filter doesn't. Reminder: use https://github.com/18F/openFEC/pull/1462 to get back the expected results from the API.

@noahmanger: when you have time, could you either check out the relevant branches and review or do a quick screenshare demo / critique?

**Comment by noahmanger** [comment]
Ok, I've got this styled enough to test:

image

I couldn't get it to be the full-width of the input + button without some annoying work, so let's see how this fares in testing and then we can refine later.

**Comment by jenniferthibault** [comment]
1000% ok with this for testing. And actually, I'd prefer to hold off on refining designs in #240 until this is tested and we have some interaction feedback. Is that ok?

**Comment by emileighoutlaw** [comment]
Is it possible to change that microcopy to read `Enter a name or committee ID` ? I think that's just a little more concise!

**Comment by jmcarp** [comment]
This is part of the consolidated typeahead feature. Re-assigning to @ccostino for review.

File: js/filters.js [hunk 1]
```diff
@@ -143,14 +143,22 @@ function TypeaheadFilter(elm) {
   Filter.call(this, elm);
 
   var key = this.$body.data('dataset');
+  var allowText = this.$body.data('allow-text') !== undefined;
   var dataset = typeahead.datasets[key];
-  this.typeaheadFilter = new typeaheadFilter.TypeaheadFilter(this.$body, dataset);
+  this.typeaheadFilter = new typeaheadFilter.TypeaheadFilter(this.$body, dataset, allowText);
```
Comment (ccostino) [inline js/filters.js]: One quick note about this - this value is going to be the result of the evaluation of `this.$body.data('allow-text') !== undefined`, not the actual value of `this.$body.data('allow-text'` itself.  I just wanted to be sure that is the intent here and not to use the actual value of the `data-allow-text` attribute.

Comment (ccostino) [inline js/filters.js]: Or, should this be expressed differently, e.g.:  `this.$body.data('allow-text') !== undefined ? this.$body.data('allow-text') : false;`?

Comment (jmcarp) [inline js/filters.js]: The intention here is that `data-allow-text` is either present or absent:

```
<div data-allow-text>text allowed</div>
<div>text not allowed</div>
```

The `allowText` variable in the javascript is meant to be a boolean flag that signals whether `data-allow-text` is set or not.

Comment (ccostino) [inline js/filters.js]: Okay, so the presence of `data-allow-text` signifies `true` regardless of what it's set to and its absence is the `false` condition for `allowText`.  Just wanted to be sure!

**Pull request merged**